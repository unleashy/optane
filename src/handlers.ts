import type { Element, Handler, HandlerResult } from "./types.ts";

const HandlerError = {
  unknown: "unknown problem",
  missingValue: "missing value for option",
  notInt: "expected an integer",
  notOneOf: "expected one of the valid values",
} as const;

/**
 * Factory for {@link Handler}.
 *
 * @param exec - The function to call when this Handler is executed.
 * @returns A properly implemented Handler.
 */
export function createHandler<T>(
  exec: (elements: Element[], i: number) => HandlerResult<T>,
): Handler<T, undefined> {
  class Impl<D extends T | undefined> implements Handler<T, D> {
    readonly #default: D;
    readonly #aliases: string[] = [];
    readonly #help: string | undefined;

    constructor(default_: D, aliases: string[], help: string | undefined) {
      this.#default = default_;
      this.#aliases = aliases;
      this.#help = help;
    }

    exec(elements: Element[], i: number): HandlerResult<T> {
      return exec(elements, i);
    }

    default(): D;
    default(value: T): Handler<T, T>;
    default(value?: T): D | Handler<T, T> {
      if (value === undefined) {
        return this.#default;
      } else {
        return new Impl(value, this.#aliases, this.#help);
      }
    }

    alias(): string[];
    alias(...names: string[]): Handler<T, D>;
    alias(...names: string[]): string[] | Handler<T, D> {
      if (names.length === 0) {
        return this.#aliases;
      } else {
        return new Impl(
          this.#default,
          [...this.#aliases, ...names],
          this.#help,
        );
      }
    }

    help(): string | undefined;
    help(text: string): Handler<T, D>;
    help(text?: string): string | Handler<T, D> | undefined {
      if (text === undefined) {
        return this.#help;
      } else {
        return new Impl(this.#default, this.#aliases, text);
      }
    }
  }

  return new Impl(undefined, [], undefined);
}

/**
 * Handler that expects a single argument and uses it as the value without any
 * further processing. Fails if no argument is present.
 */
export const string = createHandler((elements, i) => {
  let next = elements[i] as Element | undefined;
  if (next && next.type === "free") {
    return { ok: true, value: next.value, nextIndex: i + 1 };
  } else {
    return { ok: false, error: HandlerError.missingValue };
  }
});

/**
 * Handler that expects no arguments; always returns `true`. Meant for flag
 * options. If an option is not present, it defaults to `false`.
 */
export const bool = createHandler((_elements, i) => ({
  ok: true,
  value: true,
  nextIndex: i,
})).default(false);

/**
 * Handler that expects an argument and tries to parse it into an integer.
 * Fails if it canâ€™t parse its argument to an integer or if no argument is
 * present.
 */
export const int = createHandler((elements, i) => {
  let next = elements[i] as Element | undefined;
  if (next && next.type === "free") {
    if (/^\d+$/.test(next.value)) {
      return { ok: true, value: Number(next.value), nextIndex: i + 1 };
    } else {
      return { ok: false, error: HandlerError.notInt };
    }
  } else {
    return { ok: false, error: HandlerError.missingValue };
  }
});

/**
 * Handler that expects an argument and that argument must be exactly equal to
 * one of the given `strings`. Fails if its argument is not equal to any of the
 * given `strings` or if no argument is present.
 *
 * @param strings - The set of strings to expect
 */
export const oneOf = <K extends string>(...strings: [K, ...K[]]) =>
  createHandler((elements, i) => {
    let next = elements[i] as Element | undefined;
    if (next && next.type === "free") {
      if ((strings as string[]).includes(next.value)) {
        return { ok: true, value: next.value as K, nextIndex: i + 1 };
      } else {
        return { ok: false, error: HandlerError.notOneOf };
      }
    } else {
      return { ok: false, error: HandlerError.missingValue };
    }
  });

////////////////////////////////////////////////////////////////////////////////

if (import.meta.vitest) {
  const { describe, it, expect } = import.meta.vitest;

  describe("string handler", () => {
    it("uses the argument", () => {
      expect(string.exec([{ type: "free", value: "foo" }], 0)).toEqual({
        ok: true,
        value: "foo",
        nextIndex: 1,
      });
    });

    it("expects an argument", () => {
      expect(string.exec([], 0)).toEqual({
        ok: false,
        error: HandlerError.missingValue,
      });
      expect(string.exec([{ type: "option", name: "foo" }], 0)).toEqual({
        ok: false,
        error: HandlerError.missingValue,
      });
    });

    it("has default undefined", () => {
      expect(string.default()).toBe(undefined);
    });
  });

  describe("bool handler", () => {
    it("uses no arguments", () => {
      expect(bool.exec([], 0)).toEqual({
        ok: true,
        value: true,
        nextIndex: 0,
      });

      expect(bool.exec([{ type: "free", value: "foo" }], 0)).toEqual({
        ok: true,
        value: true,
        nextIndex: 0,
      });
    });

    it("has default false", () => {
      expect(bool.default()).toBe(false);
    });
  });

  describe("int handler", () => {
    it("uses the argument", () => {
      expect(int.exec([{ type: "free", value: "1234" }], 0)).toEqual({
        ok: true,
        value: 1234,
        nextIndex: 1,
      });
    });

    it("expects an argument", () => {
      expect(int.exec([], 0)).toEqual({
        ok: false,
        error: HandlerError.missingValue,
      });
      expect(int.exec([{ type: "option", name: "foo" }], 0)).toEqual({
        ok: false,
        error: HandlerError.missingValue,
      });
      expect(int.exec([{ type: "free", value: "not a number" }], 0)).toEqual({
        ok: false,
        error: HandlerError.notInt,
      });
    });

    it("has default undefined", () => {
      expect(int.default()).toBe(undefined);
    });
  });

  describe("oneOf handler", () => {
    it("uses the argument", () => {
      expect(
        oneOf("a", "b", "c").exec([{ type: "free", value: "a" }], 0),
      ).toEqual({
        ok: true,
        value: "a",
        nextIndex: 1,
      });
    });

    it("expects an argument", () => {
      let sut = oneOf("low", "high");

      expect(sut.exec([], 0)).toEqual({
        ok: false,
        error: HandlerError.missingValue,
      });
      expect(sut.exec([{ type: "option", name: "low" }], 0)).toEqual({
        ok: false,
        error: HandlerError.missingValue,
      });
      expect(sut.exec([{ type: "free", value: "?" }], 0)).toEqual({
        ok: false,
        error: HandlerError.notOneOf,
      });
      expect(sut.exec([{ type: "free", value: "medium" }], 0)).toEqual({
        ok: false,
        error: HandlerError.notOneOf,
      });
    });
  });
}
