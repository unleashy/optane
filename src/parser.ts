function snakeToCamel(snake: string): string {
  return snake.replaceAll(
    /([^-])-([^-])/g,
    (_, l: string, r: string) => `${l}${r.toUpperCase()}`,
  );
}

export type Element =
  | { type: "option"; name: string }
  | { type: "free"; value: string }
  | { type: "end-options" };

export type Result = { elements: Element[]; errors: string[] };

export function parse(argv: string[]): Result {
  let cleanArgv = argv
    .map((value) => value.trim())
    .filter((value) => value.length > 0);

  let errors: string[] = [];
  let optionsEnded = false;
  let elements: Element[] = cleanArgv.flatMap((arg): Element[] => {
    if (!optionsEnded) {
      if (arg === "--") {
        optionsEnded = true;
        return [{ type: "end-options" }];
      }

      if (arg.startsWith("--")) {
        let name = snakeToCamel(arg.slice(2));

        return [{ type: "option", name }];
      } else if (arg.startsWith("-") && arg !== "-") {
        let name = arg.slice(1);
        if (name.length > 1) {
          errors.push(
            `-${name}: short options must only use a single character`,
          );
          return [];
        }

        return [{ type: "option", name }];
      }
    }

    return [{ type: "free", value: arg }];
  });

  return { elements, errors };
}

////////////////////////////////////////////////////////////////////////////////

if (import.meta.vitest) {
  const { test, expect } = import.meta.vitest;

  test("empty args", () => {
    expect(parse([])).toMatchSnapshot();
  });

  test("one positional arg", () => {
    expect(parse(["foo"])).toMatchSnapshot();
  });

  test("multiple positional args", () => {
    expect(parse(["foo", "bar", " baz "])).toMatchSnapshot();
  });

  test("empty strings are ignored", () => {
    expect(parse(["", "a", "   ", "b", "\n", "c"])).toMatchSnapshot();
  });

  test("one short option", () => {
    expect(parse(["-a"])).toMatchSnapshot();
  });

  test("multiple short options", () => {
    expect(parse(["-x", "-y", "-z"])).toMatchSnapshot();
  });

  test("short options must be a single character", () => {
    expect(parse(["-foo"])).toMatchSnapshot();
  });

  test("one long option", () => {
    expect(parse(["--foo"])).toMatchSnapshot();
  });

  test("multiple long options", () => {
    expect(parse(["--bread", "--eggs", "--ham"])).toMatchSnapshot();
  });

  test("mixed arguments and options", () => {
    expect(
      parse(["rice", "--sausage", "french", "-f", "ries"]),
    ).toMatchSnapshot();
  });

  test("snake-case becomes camelCase", () => {
    expect(
      parse([
        "--a-b",
        "--foo-bar",
        "--is-the-thing-on",
        "--a",
        "--a--b--c",
        "--a-",
        "---b",
      ]),
    ).toMatchSnapshot();
  });

  test("- is not interpreted as an option", () => {
    expect(parse(["-"])).toMatchSnapshot();
  });

  test("-- is interpreted as the end of options", () => {
    expect(parse(["--", "-a"])).toMatchSnapshot();
  });
}
