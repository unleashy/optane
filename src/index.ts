import assert from "node:assert/strict";
import { parse } from "./parser.ts";
import * as t from "./handlers.ts";
import type { Handler, HandlerType } from "./handlers.ts";

function flatMapObject<K extends string, V, L extends string, U>(
  obj: Record<K, V>,
  f: (k: K, v: V) => Array<[L, U]>,
): Record<L, U> {
  return Object.fromEntries(
    Object.entries(obj).flatMap(([k, v]) => f(k as K, v as V)),
  ) as Record<L, U>;
}

function mapValues<K extends string, V, U>(
  obj: Record<K, V>,
  f: (v: V) => U,
): Record<K, U> {
  return flatMapObject(obj, (k, v) => [[k, f(v)]]);
}

function formatOption(name: string): string {
  return name.length === 1 ? `-${name}` : `--${name}`;
}

export type Spec = Record<string, Handler<unknown>>;

export type Options<S extends Spec> = { help: boolean } & {
  [P in keyof S]: HandlerType<S[P]> | undefined;
};

export type Result<S extends Spec> = {
  /**
   * The parsed options, as defined in your specification. Also includes the
   * `help` boolean option.
   */
  options: Options<S>;

  /** Any remaining positional arguments. */
  args: string[];

  /** Type or parse errors produced during processing */
  errors: string[];
};

export function optane<S extends Spec>(argv: string[], spec: S): Result<S> {
  let specWithHelp = { help: t.bool().alias("h"), ...spec };
  let aliases = flatMapObject(specWithHelp, (canonicalName, handler) =>
    handler.alias().map((alias) => [alias, canonicalName]),
  );
  let getHandler = (name: string) => {
    if (name in specWithHelp) {
      return { canonicalName: name, handler: specWithHelp[name] };
    } else if (name in aliases) {
      let canonicalName = aliases[name];
      return { canonicalName, handler: specWithHelp[canonicalName] };
    } else {
      return { canonicalName: undefined, handler: undefined };
    }
  };

  let options = mapValues(specWithHelp, (h) => h.default()) as Options<Spec>;
  let args: string[] = [];

  let realArgv = argv === process.argv ? argv.slice(2) : argv;
  let { elements, errors } = parse(realArgv);

  for (let i = 0; i < elements.length; ) {
    let element = elements[i++];
    switch (element.type) {
      case "option": {
        let { canonicalName, handler } = getHandler(element.name);
        if (canonicalName && handler) {
          let result = handler.exec(elements, i);
          if (result.ok) {
            options[canonicalName] = result.value;

            i = result.nextIndex;
          } else {
            let optionName = formatOption(element.name);
            if (canonicalName !== element.name) {
              optionName += ` (alias of ${formatOption(canonicalName)})`;
            }

            errors.push(`${optionName}: ${result.error}`);
          }
        } else {
          errors.push(`${formatOption(element.name)}: unknown option`);
        }

        break;
      }

      case "free": {
        args.push(element.value);
        break;
      }

      default: {
        assert.fail(`Malformed element ${JSON.stringify(element)}`);
      }
    }
  }

  return { options: options as Options<S>, args, errors };
}

export * as t from "./handlers.ts";

////////////////////////////////////////////////////////////////////////////////

if (import.meta.vitest) {
  const { test, expect } = import.meta.vitest;

  test("empty spec", () => {
    expect(optane([], {})).toMatchSnapshot();
  });

  test("skips the first two argv if it is process.argv", () => {
    expect(optane(process.argv, {})).toMatchSnapshot();
  });

  test("positional arguments are forwarded", () => {
    expect(optane(["a", "b", "c"], {})).toMatchSnapshot();
  });

  test("help option is always present", () => {
    expect(optane(["--help"], {})).toMatchSnapshot();
    expect(optane(["-h"], {})).toMatchSnapshot();
  });

  test("basic string option", () => {
    expect(optane(["--foo", "bar"], { foo: t.string() })).toMatchSnapshot();
  });

  test("string option requires argument", () => {
    expect(optane(["--wrong"], { wrong: t.string() })).toMatchSnapshot();
    expect(
      optane(["--wrong", "--wrong"], { wrong: t.string() }),
    ).toMatchSnapshot();
  });

  test("undefined options are errors", () => {
    expect(optane(["--what", "--the", "-f"], {})).toMatchSnapshot();
  });

  test("boolean options", () => {
    expect(
      optane(["--is-cool", "--awesome"], {
        isCool: t.bool(),
        awesome: t.bool(),
        epic: t.bool(),
      }),
    ).toMatchSnapshot();
  });

  test("string option with default", () => {
    expect(optane([], { food: t.string().default("bread") })).toMatchSnapshot();
  });

  test("mixed options and arguments", () => {
    expect(
      optane(["a", "--foo", "bar", "--baz", "--bux", "something", "else"], {
        foo: t.string(),
        baz: t.bool(),
        bux: t.string(),
      }),
    ).toMatchSnapshot();
  });

  test("option with long alias", () => {
    expect(
      optane(["--other", "name"], { foo: t.string().alias("other") }),
    ).toMatchSnapshot();
  });

  test("option with short alias", () => {
    expect(
      optane(["-o", "name"], { foo: t.string().alias("o") }),
    ).toMatchSnapshot();
  });

  test("short canonical option is rendered properly for error", () => {
    expect(optane(["--foo"], { o: t.string().alias("foo") })).toMatchSnapshot();
  });

  test("int option", () => {
    const sut = (args: string[]) => optane(args, { port: t.int() });

    expect(sut(["--port", "8080"])).toMatchSnapshot();
    expect(sut(["--port", "nope"])).toMatchSnapshot();
  });
}
