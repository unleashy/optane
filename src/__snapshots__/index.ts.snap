// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`-- represents end of options 1`] = `
{
  "args": [],
  "errors": [],
  "options": {
    "foo": undefined,
    "help": false,
  },
  "spec": {
    "foo": Impl {},
    "help": Impl {},
  },
}
`;

exports[`-- represents end of options 2`] = `
{
  "args": [
    "--foo",
  ],
  "errors": [],
  "options": {
    "foo": undefined,
    "help": false,
  },
  "spec": {
    "foo": Impl {},
    "help": Impl {},
  },
}
`;

exports[`-- represents end of options 3`] = `
{
  "args": [
    "--foo",
  ],
  "errors": [
    "--foo: missing value for option",
  ],
  "options": {
    "foo": undefined,
    "help": false,
  },
  "spec": {
    "foo": Impl {},
    "help": Impl {},
  },
}
`;

exports[`basic string option 1`] = `
{
  "args": [],
  "errors": [],
  "options": {
    "foo": "bar",
    "help": false,
  },
  "spec": {
    "foo": Impl {},
    "help": Impl {},
  },
}
`;

exports[`boolean options 1`] = `
{
  "args": [],
  "errors": [],
  "options": {
    "awesome": true,
    "epic": false,
    "help": false,
    "isCool": true,
  },
  "spec": {
    "awesome": Impl {},
    "epic": Impl {},
    "help": Impl {},
    "isCool": Impl {},
  },
}
`;

exports[`curried 1`] = `
{
  "args": [
    "foo",
    "bar",
  ],
  "errors": [],
  "options": {
    "help": false,
  },
  "spec": {
    "help": Impl {},
  },
}
`;

exports[`empty spec 1`] = `
{
  "args": [],
  "errors": [],
  "options": {
    "help": false,
  },
  "spec": {
    "help": Impl {},
  },
}
`;

exports[`help option is always present 1`] = `
{
  "args": [],
  "errors": [],
  "options": {
    "help": true,
  },
  "spec": {
    "help": Impl {},
  },
}
`;

exports[`help option is always present 2`] = `
{
  "args": [],
  "errors": [],
  "options": {
    "help": true,
  },
  "spec": {
    "help": Impl {},
  },
}
`;

exports[`int option 1`] = `
{
  "args": [],
  "errors": [],
  "options": {
    "help": false,
    "port": 8080,
  },
  "spec": {
    "help": Impl {},
    "port": Impl {},
  },
}
`;

exports[`int option 2`] = `
{
  "args": [
    "nope",
  ],
  "errors": [
    "--port: expected an integer",
  ],
  "options": {
    "help": false,
    "port": undefined,
  },
  "spec": {
    "help": Impl {},
    "port": Impl {},
  },
}
`;

exports[`mixed options and arguments 1`] = `
{
  "args": [
    "a",
    "else",
  ],
  "errors": [],
  "options": {
    "baz": true,
    "bux": "something",
    "foo": "bar",
    "help": false,
  },
  "spec": {
    "baz": Impl {},
    "bux": Impl {},
    "foo": Impl {},
    "help": Impl {},
  },
}
`;

exports[`oneOf option 1`] = `
{
  "args": [],
  "errors": [],
  "options": {
    "help": false,
    "verbose": "low",
  },
  "spec": {
    "help": Impl {},
    "verbose": Impl {},
  },
}
`;

exports[`oneOf option 2`] = `
{
  "args": [],
  "errors": [],
  "options": {
    "help": false,
    "verbose": "medium",
  },
  "spec": {
    "help": Impl {},
    "verbose": Impl {},
  },
}
`;

exports[`oneOf option 3`] = `
{
  "args": [],
  "errors": [],
  "options": {
    "help": false,
    "verbose": "high",
  },
  "spec": {
    "help": Impl {},
    "verbose": Impl {},
  },
}
`;

exports[`oneOf option 4`] = `
{
  "args": [
    "what",
  ],
  "errors": [
    "--verbose: expected one of the valid values",
  ],
  "options": {
    "help": false,
    "verbose": undefined,
  },
  "spec": {
    "help": Impl {},
    "verbose": Impl {},
  },
}
`;

exports[`option with long alias 1`] = `
{
  "args": [],
  "errors": [],
  "options": {
    "foo": "name",
    "help": false,
  },
  "spec": {
    "foo": Impl {},
    "help": Impl {},
  },
}
`;

exports[`option with negative int argument 1`] = `
{
  "args": [],
  "errors": [],
  "options": {
    "foo": -1,
    "help": false,
  },
  "spec": {
    "foo": Impl {},
    "help": Impl {},
  },
}
`;

exports[`option with short alias 1`] = `
{
  "args": [],
  "errors": [],
  "options": {
    "foo": "name",
    "help": false,
  },
  "spec": {
    "foo": Impl {},
    "help": Impl {},
  },
}
`;

exports[`positional arguments are forwarded 1`] = `
{
  "args": [
    "a",
    "b",
    "c",
  ],
  "errors": [],
  "options": {
    "help": false,
  },
  "spec": {
    "help": Impl {},
  },
}
`;

exports[`short canonical option is rendered properly for error 1`] = `
{
  "args": [],
  "errors": [
    "--foo (alias of -o): missing value for option",
  ],
  "options": {
    "help": false,
    "o": undefined,
  },
  "spec": {
    "help": Impl {},
    "o": Impl {},
  },
}
`;

exports[`skips the first two argv if it is process.argv 1`] = `
{
  "args": [],
  "errors": [],
  "options": {
    "help": false,
  },
  "spec": {
    "help": Impl {},
  },
}
`;

exports[`string option requires argument 1`] = `
{
  "args": [],
  "errors": [
    "--wrong: missing value for option",
  ],
  "options": {
    "help": false,
    "wrong": undefined,
  },
  "spec": {
    "help": Impl {},
    "wrong": Impl {},
  },
}
`;

exports[`string option requires argument 2`] = `
{
  "args": [],
  "errors": [
    "--wrong: missing value for option",
    "--wrong: missing value for option",
  ],
  "options": {
    "help": false,
    "wrong": undefined,
  },
  "spec": {
    "help": Impl {},
    "wrong": Impl {},
  },
}
`;

exports[`string option with default 1`] = `
{
  "args": [],
  "errors": [],
  "options": {
    "food": "bread",
    "help": false,
  },
  "spec": {
    "food": Impl {},
    "help": Impl {},
  },
}
`;

exports[`undefined options are errors 1`] = `
{
  "args": [],
  "errors": [
    "--what: unknown option",
    "--the: unknown option",
    "-f: unknown option",
  ],
  "options": {
    "help": false,
  },
  "spec": {
    "help": Impl {},
  },
}
`;
